### ISSUE TITLE
[MU4 Issue] Design question regarding W/Q shortcuts + Shift+W/Q

### ISSUE THREAD

**Issue by worldwideweary** [musescore/MuseScore#10680]
More of a design question rather than a strict issue:

Question: Why the move from Q/W not affecting [most recent input] during Note Entry (like in 3.x) but instead merely acting as equivalent functions of pressing the duration toggles (with the specification of doubling/halving)? It seems to me that functionality is lost here, and if it's for the sake of uniformity, to make matters more interesting, Shift+W/Q still applies to the [most recent input] instead of merely acting again as the toggle for dotting. It feels funny that Shift+Q/W shortcuts are post-application while Q/W are pre-toggle shortcuts, but that's just my impression. I'm sure others will be able to get used to it either way.

In other words: Is it intended for the Shift+W/Q shortcuts to end up becoming toggles also, or will those stay as post-application in conjunction with Q/W being pre-toggle? Or is this just testing ground for a change in Q/W functionality? Thanks

**Comment by jeetee** [comment]
I've been part of the design discussion in the devchat around this change. It was brought along because currently Q/W was tied to 3 different actions and a desire to unify their usage.

To me, it always was weird that (shift+)Q/W during note entry input worked post-entry when no other duration related command does (normal duration selection, dots are all pre-entry and accidentals are also pre-entry). So the consensus here was that (Shift+)Q/W should act in a similar way as other duration related commands:
* During note entry: affect the next to be entered item
* Outside of note entry: affect the selected items

So yes, Shift+Q/W should also be working pre-entry according to me during note-entry. I just assume it not being so currently an oversight in the initial implementation fix. Let's keep this issue open as a reminder task for @RomanPudashkin

**Comment by worldwideweary** [comment]
Ok, thanks for the response.
For the record, if I remember correctly, [Shift+Q/W] was implemented prior to the change from post-application of accidentals into toggle-only, so it wasn't always the only post-entry command, along with of course the 3.x behavior of Q/W.  (and after forum conversations I figured what the hell and also made separate shortcuts for post-entry behavior for users who had acclimated themselves to the 1.x-3.4orsoish behavior).

**Comment by MarcSabatella** [comment]
Pretty much the only reason I ever use Q or W is as error correction - when I forget to select the correct duration.  I would hate to lose this, and I'd essentially never use Q or Q otherwise.  So to me post makes way more.  But, I'd *prefer* to see a more direct way to post-apply durations, rather than kludging it with Q & W.  If I want to retroactively apply quarter note, I want to press 5 ! (or whatever the shortcut is, that's actually not a bad one for people coming Linux) and be done with it, not have to think bout whether to use Q or W and how many presses.

So, I'd withdraw my objection to the Q & W if someone were to consider https://github.com/musescore/MuseScore/pull/5505 or something similar :-)

**Comment by MarcSabatella** [comment]
See in particular my comment in https://github.com/musescore/MuseScore/pull/5505#issuecomment-644248247 - it seems retroactively changing duration is considered an essential part of the workflow for many ex-Sibelius folks.

**Comment by worldwideweary** [comment]
For what it's worth, I also almost strictly only use (shift+)Q/W for post-correction and never preparation. For preparation, I use the duration toggles + the augmentation commands. On occasion I might 'ride' input in a way that allows me to be lazy in using Q/W for a pseudo-correction rather than needing to be precise in the pre-toggle. For example, I can use Q/W with my left hand  quickly, or more easily, than having to press numbers if i need to shift between quarter and eighth back and forth, and because of that ease in hand motion from home-row position, I might use it for that very reason as an aftermath, even though it's not 'designed with that in mind'.

**Comment by bkunda** [comment]
Good observation @worldwideweary. @MarcSabatella I also wouldn't advocate for removing functionality in this case, particularly if there is enough of a user need for it. 

FWIW, I had come up with an alternative way to streamline the current `increase/decrease active duration` and `double/halve duration` shortcuts so that they work in the same way across both TAB and standard staves. I'm sure it's already known to people here, but it might be worth reiterating that having W and Q (or something equivalent) affect the _next note_ is important for TAB users, so that they can cycle through note durations (the number keys being otherwise occupied with inserting the actual TAB numbers themselves). 

I actually quite liked @jeetee's solution because it made the use of Q and W predictable in different states, but I hadn't considered that it might stop people from correcting their previously-entered notes. So I'd like to put an alternative solution out there that might preserve this functionality:

* Increase/Decrease Active Duration (Q / W) works in note input mode only (on both standard and TAB staves) and affects the next note to be entered
* Double/Halve Duration (I'd suggest: Alt+Q / Alt+W) works on selected notes/rests only (on both standard and TAB staves) and can be used to correct the previously-entered note

This latter command would also align better with the current Shift+Q/Shift+Q which, contrary to its label (`increase/decrease duration dotted`), actually seems to affect the previously-entered note rather than the next note. 

Let me know if this might solve the problem. Also @MarcSabatella I'm not quite sure I have my head around #5505 so maybe we can chat about it sometime so I can understand it better :-) 

@RomanPudashkin fyi

**Comment by bkunda** [comment]
Per @RomanPudashkin's pull request, the fix for this has already been merged for now. But of course we can still look for ways to recover this functionality – perhaps an even smarter solution than what we've already found. 

@MarcSabatella feel free to have a chat with me separately about [#5505](https://github.com/musescore/MuseScore/pull/5505) sometime :-)

**Comment by MarcSabatella** [comment]
I have nothing against making Q & W work differently in tab, given that we already make the number keys themselves work differently.  But, to be clear: it's not *necessary* to have Q & W in tab to have a good keyboard workflow, and I suspect the majority of users won't ever use it.  Pressing Q two or three times to change duration from long to short is not faster than press Shift+3 or whatever, and it requires much more "think time" to figure out how many presses will be required.  Maybe a handful of users will do this, but I assume most would prefer to simply use Shift+number to set duration.  Assuming they are using the keyboard at all.

But just as when we changed the accidental behavior to also be pre-input, there is no denying that pre-input its more consistent and logical in many ways, and new users probably will like it just fine.  Existing ones will complain long and hard about it being "change for the sake of change" just as they did with the accidentals, but I'm OK with the change *if* we provide a viable alternative, which we didn't at first for accidentals.  Later we added a partial workaround - keyboard commands to retroactively apply each accidental - but it is not ideal since it requires learning five new commands (one for each accidental) and it's only doable via keyboard, not toolbar.

The point of my PR is to solve the "how to retroactively apply accidentals or durations" problem more generally.  A single command that you can use to take whatever you just pressed on the toolbar and apply it to the most-recently-entered note (in note input mode).  So, if you forgot to change to sixteenth note *before* entering the note, it's not too late - press 3 after, then the new command to apply it to the note you just entered.  Or if you forgot to press the sharp button before entering the note, press the sharp button after, then the new command to apply it.  The command simply takes the current "input state" (set of toolbar buttons that are pressed in anticipation of entering the next note) and applies them to the *currently selected* note.(which is normally the one you just entered), simple as that.

As it is in my PR, this a keyboard command only, but it could also be worth adding to the note input toolbar, either by default or as an option for customization.  That way mouse users can also benefit.

To me it's a better solution than Q and W ever was.  And indeed, Alt shortcuts are iffy.

**Comment by bkunda** [comment]
Thanks for this explainer, Marc. It makes much more sense to me now to understand this as a separate command that applies the selected duration/accidental to the note you’ve just entered. I agree with you that there is some potential here for possible UI interaction as well, so I have logged this on my list of possible note input/key command improvements for when we discuss our design goals post MS4.0.
